File: dcp-solutions/2024-09-16/solution.ex
```elixir
defmodule Decode Ways do
  @spec decode_ways(String.t()) :: non_neg_integer
  def decode_ways(message) when byte_size(message) == 0, do: 0
  def decode_ways("0"), do: 0
  def decode_ways(message), do: Enum.reduce(1..byte_size(message)-1, 0, fn i, acc ->
    # If the first two characters are not '00', we can always add them to the current number of ways.
    if String.slice(message, 0, 2) != "00", do: acc + 1
    # Otherwise, we need to check whether the next character is '0' or '1'.
    else Enum.at(message, i+1) == "0" ? acc : acc + 2 end)
end
```

File: dcp-solutions/2024-09-16/solution.go
```go
package main

import (
    "fmt"
)

func numDecodings(s string) int {
    if len(s) == 0 {
        return 0
    }
    
    n := len(s)
    
    // dp[i] represents the number of ways to decode s[:i]
    dp := make([]int, n+1)
    dp[0] = 1
    dp[1] = 1
    
    for i := 2; i <= n; i++ {
        if s[i-1] != '0' {
            // If the last digit can be decoded independently
            dp[i] += dp[i-1]
        }
        
        // If the last two digits together can be decoded
        if i >= 2 && (s[i-2] == '1' || (s[i-2] == '2' && s[i-1] <= '6')) {
            dp[i] += dp[i-2]
        }
    }
    
    return dp[n]
}

func main() {
    
}